### PR TITLE
Implement download_file in widget driver

### PR DESCRIPTION
**Pull request by weeman1337**
Implement the widget API `download_file` in the widget driver.

Needs https://github.com/matrix-org/matrix-widget-api/pull/99

## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [x] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

File: test/stores/widgets/StopGapWidgetDriver-test.ts
Comment (HarHarLinks) [inline test/stores/widgets/StopGapWidgetDriver-test.ts]: wouldn't this be
```suggestion
                    return "https://example.com/_matrix/media/v3/download/example.com/test_file";
                }

                return null;
            });

            fetchMockJest.get("https://example.com/_matrix/media/v3/download/example.com/test_file", "test contents");
```

it might not be obvious why this calls the deprecated API, but maintainers will know better than me.

**Review by HarHarLinks — CHANGES_REQUESTED** [review]

File: test/stores/widgets/StopGapWidgetDriver-test.ts
Comment (weeman1337) [reply test/stores/widgets/StopGapWidgetDriver-test.ts]: > wouldn't this be

Done

> it might not be obvious why this calls the deprecated API, but maintainers will know better than me.

The implementation relies on [`Media`](https://github.com/matrix-org/matrix-react-sdk/blob/ec829435bfe4ab8010740fc6884571e23f6e652e/src/customisations/Media.ts#L36) . Not sure if the test should mock it away or not :shrug:

**Comment by florianduros** [comment]
Hi! Can you first of all fix the lint and ts errors?
Thanks!

**Comment by weeman1337** [comment]
> Hi! Can you first of all fix the lint and ts errors? Thanks!

The „lint“ errors are actually type check errors caused by this PR relying on https://github.com/matrix-org/matrix-widget-api/pull/99

**Comment by HarHarLinks** [comment]
here's described how to test everything together: https://github.com/matrix-org/matrix-widget-api/pull/99#issuecomment-2314962447

**Comment by toger5** [comment]
> The „lint“ errors are actually type check errors caused by this PR relying on https://github.com/matrix-org/matrix-widget-api/pull/99

I think you should be able to update the package.json to use your widget api pr for the review and revert the change once the widget api is merged.
Than the two PR's can be reviewed simultaneously with both being checked by the linter already.

To be more specific:
Change:
https://github.com/matrix-org/matrix-react-sdk/blob/develop/package.json#L118
to
`"matrix-widget-api: "github:nordeck/matrix-widget-api#download-file"` (I think)

**Comment by florianduros** [comment]
The change in the package.json made it worse

**Comment by HarHarLinks** [comment]
> The change in the package.json made it worse

Hello @florianduros, could you please advise us what you need us to do to get this reviewed? Like weeman said, the lint errors are due to the wrong dependency version. How should we fix it?

**Comment by florianduros** [comment]
I'll review it in the afternoon. Why not waiting for the https://github.com/matrix-org/matrix-widget-api/pull/99 PR to be released to ask for a review? It can't be merged without it.

**Comment by HarHarLinks** [comment]
> Why not waiting for the [matrix-org/matrix-widget-api#99](https://github.com/matrix-org/matrix-widget-api/pull/99) PR to be released to ask for a review? It can't be merged without it.

Code for both (and also a widget) is already done so we could test it end to end. We're asking for reviews of both PRs that target element repos. It seems to us to make sense to review both PRs at the same time.

---

Some info maybe not obvious from the title and OP: this is a (partial) implementation of [MSC4039](https://github.com/matrix-org/matrix-spec-proposals/pull/4039#issuecomment-2311853691).

**Comment by dbkr** [comment]
I've merged the widgets api PR and released it as 1.9.0 so you'll need tov change the dependency here (or allow edits so we can).

**Review by dbkr — APPROVED** [review]
Similar to the widget-api PR, I have a few questions about the MSC but this seems like a god and accurate impl of what it currently says, so I'd say go with it.

**Pull request merged**
